src/main/java/duke/util/TaskList.java

```diff
@@ -0,0 +1,133 @@
+package duke.util;
+
+import java.util.ArrayList;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import duke.date.Date;
+import duke.exception.NoSuchTaskException;
+import duke.task.DatedTask;
+import duke.task.Task;
+
+/** A wrapper class that holds the list of Tasks. */
+public class TaskList {
+    /** The actual list holding the Tasks. */
+    private ArrayList<Task> list;
+
+    /** Default TaskList constructor. */
+    public TaskList() {
+        list = new ArrayList<>();
+    }
+
+    /**
+     * TaskList constructor. Used for constructing the list from storage.
+     *
+     * @param list The ArrayList of Tasks.
+     */
+    public TaskList(ArrayList<Task> list) {
+        this.list = list;
+    }
+
+    /**
+     * Adds a Task to the list.
+     *
+     * @param task The Task object.
+     */
+    public void add(Task task) {
+        list.add(task);
+    }
+
+    /**
+     * Returns the size of the list.
+     *
+     * @return Size of the list.
+     */
+    public int size() {
+        return list.size();
+    }
+
+    /**
+     * Returns the Task at the specified index in the list.
+     *
+     * @param index The index of Task to be obtained.
+     * @return The Task at the specified index.
+     * @throws NoSuchTaskException If the index exceeds the size of the list.
+     */
+    public Task getTask(int index) throws NoSuchTaskException {
+        try {
+            return list.get(index);
+        } catch (IndexOutOfBoundsException e) {
+            throw new NoSuchTaskException();
+        }
+    }
+
+    /**
+     * Deletes the Task from the specified index and returns the deleted Task.
+     *
+     * @param index The index of Task to be deleted.
+     * @return The Task deleted at the specified index.
+     * @throws NoSuchTaskException If the index exceeds the size of the list.
+     */
+    public Task deleteTask(int index) throws NoSuchTaskException {
+        try {
+            return list.remove(index);
+        } catch (IndexOutOfBoundsException e) {
+            throw new NoSuchTaskException();
+        }
+
+    }
+
+    /**
+     * Returns an array of string representation of the Tasks in the list, numbered from 1.
+     *
+     * @return An array of string representation of the Tasks in the list, numbered from 1.
+     */
+    public String[] toStringArray() {
+        String[] copy = list.stream().map(Object::toString).toArray(String[]::new);
+        for (int i = 0; i < size(); i++) {
+            copy[i] = (i + 1) + "." + copy[i];
+        }
```

Comment:
Nice! Unfortunate we don't have combine for Java streams or you could use stream operations for the indexing too.

Reply:
My feelings exactly... the code would've looked neater without the for loop, but it is Java, and with it comes clumsy syntax :(